include/babeltrace/so-info.h

```diff
@@ -0,0 +1,209 @@
+	uint64_t high_addr;
+	/* Size of exec address space. */
+	uint64_t memsz;
+	/* Paths to ELF and DWARF files. */
+	char *elf_path;
+	char *dwarf_path;
+	/* libelf and libdw objects representing the files. */
+	Elf *elf_file;
+	Dwarf *dwarf_info;
+	/* Optional build ID info. */
+	uint8_t *build_id;
+	size_t build_id_len;
+	/* Optional debug link info. */
+	char *dbg_link_filename;
+	uint32_t dbg_link_crc;
+	/* FDs to ELF and DWARF files. */
+	int elf_fd;
+	int dwarf_fd;
+	/* Denotes whether the executable is position independent code. */
+	uint8_t is_pic : 1;
+	/* Denotes whether the SO only has ELF symbols and no DWARF info. */
+	uint8_t is_elf_only : 1;
+};
+
+struct source_location {
+	long long unsigned int line_no;
+	char *filename;
+};
+
+/**
+ * Initializes the so_info framework. Call this before calling
+ * anything else.
+ *
+ * @returns		0 on success, -1 on failure
+ */
+BT_HIDDEN
+int so_info_init(void);
+
+/**
+ * Sets the so_info framework's debug directory, in which to look for
+ * separate debug info.
+ *
+ * @param path	Path to the debug info directory
+ */
+BT_HIDDEN
+void so_info_set_debug_dir(const char *path);
+
+/**
+ * Instantiate a structure representing an ELF executable, possibly
+ * with DWARF info, located at the given path.
+ *
+ * @param path		Path to the ELF file
+ * @param low_addr	Base address of the executable
+ * @param memsz	In-memory size of the executable
+ * @returns		Pointer to the new so_info on success,
+ *			NULL on failure.
+ */
+BT_HIDDEN
+struct so_info *so_info_create(const char *path, uint64_t low_addr,
+		uint64_t memsz);
+
+/**
+ * Destroy the given so_info instance
+ *
+ * @param so	so_info instance to destroy
+ */
+BT_HIDDEN
+void so_info_destroy(struct so_info *so);
+
+/**
+ * Sets the build ID information for a given so_info instance.
+ *
+ * @param so		The so_info instance for which to set
+ *			the build ID
+ * @param build_id	Array of bytes containing the actual ID
+ * @param build_id_len	Length in bytes of the build_id
+ * @returns		0 on success, -1 on failure
+ */
+BT_HIDDEN
+int so_info_set_build_id(struct so_info *so, uint8_t *build_id,
+		size_t build_id_len);
+
+/**
+ * Sets the debug link information for a given so_info instance.
+ *
+ * @param so		The so_info instance for which to set
+ *			the debug link
+ * @param filename	Name of the separate debug info file
+ * @param crc		Checksum for the debug info file
+ * @returns		0 on success, -1 on failure
+ */
+BT_HIDDEN
+int so_info_set_debug_link(struct so_info *so, char *filename, uint32_t crc);
+
+/**
+ * Returns whether or not the given SO info \p so contains the address
+ * \p addr.
+ *
+ * @param so		so_info instance
+ * @param addr		Address to lookup
+ * @returns		1 if \p so contains \p addr, 0 if it does not,
+ *			-1 on failure
+ */
+static inline
+int so_info_has_address(struct so_info *so, uint64_t addr)
+{
+	if (!so) {
+		return -1;
+	}
+
+	return addr >= so->low_addr && addr <= so->high_addr;
```

Comment:
Is the high_addr really inclusive?
I see `so->high_addr = so->low_addr + so->memsz;` in so-info.c:136 which makes it seem like it is an exclusive upper-bound.